Implement Zipkin 2 JSON Sender

keitwb opened the pull request:
The existing ZipkinSender class only supported the Thrift v1 protocol,
but the new Zipkin2Sender supports both v1 and v2 JSON.  Otherwise,
functionality is mostly the same.

There is some conceptually duplicate code in the Sender classes due to
the use of different classes between Zipkin 1/2 but it seemed overly
complicated to adapt them to a common adapter interface.

Fixes #383.

mpetazzoni commented on jaeger-zipkin/src/main/java/io/jaegertracing/senders/zipkin/V2SpanConverter.java:
```diff
@@ -0,0 +1,193 @@
+
+  public static zipkin2.Span convertSpan(Span span) {
+    Tracer tracer = span.getTracer();
+    Endpoint host = zipkin2.Endpoint.newBuilder()
```
To avoid confusion I would remove the `import zipkin2.Endpoint` and use `zipkin2.Endpoint host = ...` here.

mpetazzoni commented on jaeger-zipkin/src/main/java/io/jaegertracing/senders/zipkin/ConverterUtil.java:
```diff
@@ -0,0 +1,39 @@
+
+  static boolean isRpc(Span span) {
+    Object spanKindValue = span.getTags().get(Tags.SPAN_KIND.getKey());
+    return Tags.SPAN_KIND_CLIENT.equals(spanKindValue) || Tags.SPAN_KIND_SERVER.equals(spanKindValue);
```
I know you're just moving code around here, but this could simply be `return isRpcServer(span) || isRpcClient(span)`. Also clean up the superfluous empty newlines here and at the end of the class.

mpetazzoni commented on jaeger-zipkin/src/main/java/io/jaegertracing/senders/zipkin/V2SpanConverter.java:
```diff
@@ -0,0 +1,193 @@
+    Tracer tracer = span.getTracer();
+    Endpoint host = zipkin2.Endpoint.newBuilder()
+        .ip(convertIp(tracer.getIpv4()))
+        .port(0)
```
You can probably just omit that. Might be good to add a comment saying that we don't have a local endpoint port to put in at all.

mpetazzoni commented on jaeger-zipkin/src/main/java/io/jaegertracing/senders/zipkin/V2SpanConverter.java:
```diff
@@ -0,0 +1,193 @@
+  }
+
+  private static zipkin2.Span.Kind convertKind(Object kind) {
+    if (kind == null) {
```
If you reverse your equals calls you don't need that initial null check and have a single location where you return null from this method

mpetazzoni commented on jaeger-zipkin/src/main/java/io/jaegertracing/senders/zipkin/V2SpanConverter.java:
```diff
@@ -0,0 +1,193 @@
+  }
+
+  private static InetAddress convertIp(Integer ip) {
+    byte[] bytes = ByteBuffer.allocate(4).putInt(ip).array();
```
Can `ip` be null here? If so, you probably need to handle that. If not, then you can change the method parameter to be the unboxed `int`

jpkrohling commented:
@keitwb, thank you for this PR and sorry for the delay in reviewing it. I've been AFK for most of last week and will be for most of this week again.

Perhaps @pavolloffay would have some time this week to review, otherwise, I'll review it next week.

pavolloffay commented:
Based on https://github.com/jaegertracing/jaeger-client-java/issues/383#issuecomment-380271254 I think it would be better to have an adapter for zipkin reporter. It's more flexible and allows us
* use any zipkin supported sender: okhttp, urlconnection, kafka...
* leverage zipkin reporters
*  less code to maintain?

keitwb commented:
@pavolloffay I added an abstract base class for the two Zipkin senders so that there is quite a bit less duplication now.  This always allowed using any kind of Zipkin sender (http clients, kafka, etc.), and I don't understand what you mean about a Zipkin reporter adapter.

pavolloffay commented:
I mean instead of implementing sender interface e.g. `ZipkinSender` we could implement `ZipkinReporter` 

```java
class ZipkinReporter implements io.jaegertracing.Reporter {
   private io.zipkin.Reporter delegate;

   public ZipkinReporter(io.zipkin.Reporter reporter) {
     this.delegate = reporter;
   {

   public report(io.jaegertracing.Span jaegerSpan) {
       io.zipkin.Span ipkinSpan  = convertToZipkin(span);
       delegate.report(span);
   }
}
```

pavolloffay commented:
Something like https://github.com/jaegertracing/jaeger-client-java/commit/431a558d6b939031b062bb67f5e484298f6255ab. 

But note that it deals only with jsonv1 and jsonv2. It should be also possible to add thrift.

One disadvantage is that flush is not supporter, but we could  bind to zipkin's `AsyncReporter` (instead to `zipkin.Reporter`) which implements flush.

keitwb commented:
@pavolloffay ok, yes I have started reimplementing this using reporters and it seems much cleaner and makes much better reuse of zipkin code instead of reinventing the wheel in this lib.  My main question at this point is whether there is anything special about the use of the older `com.twitter.zipkin.thriftjava.Span` class vs. just converting the jaeger spans to `zipkin.Span` and using the [Thrift encoding](https://github.com/openzipkin/zipkin-reporter-java/blob/release-1.1.1/core/src/main/java/zipkin/reporter/Encoder.java#L32) from the Zipkin v1 reporter.

pavolloffay commented:
I think those two should be compatible.

keitwb commented:
@pavolloffay  ok, I made an initial attempt at reusing Zipkin reporters as much as possible.  It is a bit tricky since there are Zipkin 1 and 2 reporters with a different Java interface, but I think I have simplified it as much as possible.  

This does break existing users of this lib since it totally removes ZipkinSender from this repo, but I am unsure if that is important or not.

keitwb commented on jaeger-zipkin/build.gradle:
```diff
@@ -1,12 +1,16 @@
 dependencies {
-    compile group: 'io.zipkin.reporter', name: 'zipkin-reporter', version: '1.1.2'
-    compile group: 'io.zipkin.reporter', name: 'zipkin-sender-urlconnection', version: '1.1.2'
+    compile group: 'io.zipkin.reporter', name: 'zipkin-reporter', version: '1.0.4'
```
I had to downgrade this because of some nasty classpath conflicts between the version of Zipkin that 1.1.2 and 2.6.0 of the reporter were using.  Is there some way to pin the version of Zipkin used by dependencies of this module?

pavolloffay replied:
What If you exclude the old zipkin(2.2.1) from reporter 1.1.2?

keitwb replied:
I'll see if it is still an issue after I revert the ZipkinSender and only implement the V2 Reporter.

keitwb commented on jaeger-zipkin/src/main/java/io/jaegertracing/senders/zipkin/V2SpanConverter.java:
```diff
@@ -0,0 +1,187 @@
+        endLabel = zipkin.Constants.CLIENT_RECV;
+      }
+
+      builder.addAnnotation(span.getStart(), startLabel);
```
Is this necessary in v2?

pavolloffay replied:
I don't think so 

cs time - is kind=client and start time
cr time - is kind=client starttime+duration

the same for server side, but the kind is server

keitwb commented:
@pavolloffay How does it look now?

pavolloffay commented:
I think it's better to keep the zipkin sender and just deprecate it.

pavolloffay commented on jaeger-zipkin/src/test/java/io/jaegertracing/senders/zipkin/V2SpanConverterTest.java:
```diff
@@ -0,0 +1,310 @@
+  @Before
+  public void setUp() {
+    tracer =
+        new Tracer.Builder("test-service-name", new InMemoryReporter(), new ConstSampler(true))
```
please do not use deprecated APIs

keitwb commented:
@pavolloffay ok, I reverted the original ZipkinSender and rearranged things a bit.  I think I figured out how to exclude Zipkin 2.2.1 from polluting the classpath.  Let me know what you think.

pavolloffay commented:
@keitwb could you please keep `ThriftSpanConverter` in the old package? Just keep the changes minimal

pavolloffay commented on jaeger-zipkin/README.md:
```diff
@@ -21,12 +21,39 @@ tracer = new Tracer.Builder(serviceName, reporter, sampler)
+import zipkin2.reporter.AsyncReporter;
+import zipkin2.reporter.urlconnection.URLConnectionSender;
+
+reporter = new Zipkin2Reporter(
```
nit: could you please just rename it to `ZipkinV2Reporter`?

pavolloffay commented on jaeger-zipkin/src/main/java/io/jaegertracing/zipkin/V2SpanConverter.java:
```diff
@@ -0,0 +1,187 @@
+        endLabel = zipkin.Constants.CLIENT_RECV;
+      }
+
+      builder.addAnnotation(span.getStart(), startLabel);
```
From the previous review: these are not needed in V2

pavolloffay commented:
@keitwb the PR looks good, I had just a couple of minor comments.

pavolloffay commented:
@keitwb sorry It also needs a rebase due to my merges today :/

keitwb commented on jaeger-zipkin/src/main/java/io/jaegertracing/zipkin/V2SpanConverter.java:
```diff
@@ -0,0 +1,187 @@
+        componentName = span.getTracer().getServiceName();
+      }
+
+      builder.putTag(zipkin.Constants.LOCAL_COMPONENT, componentName);
```
What about this one?  In Zipkin v2 spans, the local endpoint field contains the service name so this seems redundant to add this special tag.

I'm going to remove it for now but let me know if you think it should remain.

keitwb commented:
@pavolloffay ok, I moved the Thrift converter back to its original location and rebased.  Also, I removed some logic for the local component special tag since it was in the local endpoint object already (see comment).  I put the new stuff as a separate commit so you can see it clearly.  Thanks!

keitwb commented:
Ok, I did the same thing the Zipkin reporter tests do and it seemed to fix it.

pavolloffay commented:
@keitwb great work! Thanks for the PR

I will wait for 24h in case anybody else wants to review

jpkrohling commented on jaeger-zipkin/build.gradle:
```diff
@@ -1,12 +1,23 @@
+        exclude group:'io.zipkin.java'
+    }
+    compile(group: 'io.zipkin.reporter', name: 'zipkin-sender-urlconnection', version: '1.1.2') {
+        exclude group:'io.zipkin.zipkin2'
```
Is this needed, if you are declaring it as a direct dependency of your project?

keitwb replied:
Tested it, and you are right, it doesn't appear to be necessary due to the explicit dependency.

jpkrohling commented on jaeger-zipkin/src/main/java/io/jaegertracing/zipkin/V2SpanConverter.java:
```diff
@@ -0,0 +1,161 @@
+    } else if (Tags.SPAN_KIND_PRODUCER.equals(kind)) {
+      return zipkin2.Span.Kind.PRODUCER;
+    } else {
+      return null;
```
Is this an expected case? If not, a log statement should be produced, like, "Could not identify the span kind during conversion. Skipping."

pavolloffay replied:
IIRC internal spans has a null kind.

pavolloffay replied:
yes they have kind set to  `null`

jpkrohling commented on jaeger-zipkin/src/main/java/io/jaegertracing/zipkin/V2SpanConverter.java:
```diff
@@ -0,0 +1,161 @@
+    try {
+      return InetAddress.getByAddress(bytes);
+    } catch (UnknownHostException e) {
+      return null;
```
IIRC, this *should* never happen but would be worth adding a log statement here.

jpkrohling commented on jaeger-zipkin/src/test/java/io/jaegertracing/zipkin/V2SpanConverterTest.java:
```diff
@@ -0,0 +1,266 @@
+  Tracer tracer;
+
+  @Before
+  public void setUp() {
```
`s/setUp/setup`

pavolloffay commented:
I am just linking a similar PR in go https://github.com/jaegertracing/jaeger-client-go/pull/310. I have updated the branch and I will merge on green
